### ISSUE TITLE
EPIC: Migrate videos poll to Mehestan algorithm

### ISSUE THREAD

**Issue by lfaucon** [tournesol-app/tournesol#950]
A first problem identified so far is the scaling of scores that is very different to Lichavi's. However this is only a small problem and we can easy scale manually somehow or adapt the visualizations on the frontend.

The Lichavi's algorithm fails to correctly estimate individual preferences and it appears that all individual contributors are very close to the global ranking, which shouldn't be the case.

Before making this change, it would be best to test on staging with data from prod, and check that rankings look roughly correct.

If more tasks are needed to achieve the migration, let's add them on this ticket.

**Comment by amatissart** [comment]
I ran Mehestan with data from a production backup dated today (2022-05-26).
The job run on my local VM (with a single process) in 12min46s on the branch [`mehestan-tournesol-score`](https://github.com/tournesol-app/tournesol/compare/3999c55)
Even though the CPUs may not be as performant, we can expect the job to run significantly faster on the production VM thanks to the criteria parallelization among multiple processes.

<details>
<summary> Click to expand ML logs</summary>

```
May 26 20:19:12 tournesol systemd[1]: Starting Train ML...
May 26 20:19:14 tournesol bash[21329]: NumExpr defaulting to 2 threads.
May 26 20:19:14 tournesol bash[21329]: Mehestan for poll 'videos': Start
May 26 20:19:14 tournesol bash[21337]: Mehestan for poll 'videos': computing scores for crit 'largely_recommended'
May 26 20:19:22 tournesol bash[21337]: Individual scores computed for crit 'largely_recommended'
May 26 20:19:43 tournesol bash[21337]: Computing scaling for 2940 non_supertrusted, based on 72 supertrusted
May 26 20:21:01 tournesol bash[21337]: Mehestan for poll 'videos': scores computed for crit 'largely_recommended' and mode 'default'
May 26 20:21:05 tournesol bash[21337]: Mehestan for poll 'videos': scores computed for crit 'largely_recommended' and mode 'all_equal'
May 26 20:21:07 tournesol bash[21337]: Mehestan for poll 'videos': scores computed for crit 'largely_recommended' and mode 'trusted_only'
May 26 20:21:08 tournesol bash[21337]: Mehestan for poll 'videos': done with crit 'largely_recommended'
May 26 20:21:08 tournesol bash[21337]: Mehestan for poll 'videos': computing scores for crit 'reliability'
May 26 20:21:14 tournesol bash[21337]: Individual scores computed for crit 'reliability'
May 26 20:21:35 tournesol bash[21337]: Computing scaling for 2940 non_supertrusted, based on 72 supertrusted
May 26 20:22:16 tournesol bash[21337]: Mehestan for poll 'videos': scores computed for crit 'reliability' and mode 'default'
May 26 20:22:19 tournesol bash[21337]: Mehestan for poll 'videos': scores computed for crit 'reliability' and mode 'all_equal'
May 26 20:22:21 tournesol bash[21337]: Mehestan for poll 'videos': scores computed for crit 'reliability' and mode 'trusted_only'
May 26 20:22:22 tournesol bash[21337]: Mehestan for poll 'videos': done with crit 'reliability'
May 26 20:22:22 tournesol bash[21337]: Mehestan for poll 'videos': computing scores for crit 'pedagogy'
May 26 20:22:26 tournesol bash[21337]: Individual scores computed for crit 'pedagogy'
May 26 20:22:40 tournesol bash[21337]: Computing scaling for 2940 non_supertrusted, based on 72 supertrusted
May 26 20:23:24 tournesol bash[21337]: Mehestan for poll 'videos': scores computed for crit 'pedagogy' and mode 'default'
May 26 20:23:27 tournesol bash[21337]: Mehestan for poll 'videos': scores computed for crit 'pedagogy' and mode 'all_equal'
May 26 20:23:29 tournesol bash[21337]: Mehestan for poll 'videos': scores computed for crit 'pedagogy' and mode 'trusted_only'
May 26 20:23:30 tournesol bash[21337]: Mehestan for poll 'videos': done with crit 'pedagogy'
May 26 20:23:30 tournesol bash[21337]: Mehestan for poll 'videos': computing scores for crit 'importance'
May 26 20:23:35 tournesol bash[21337]: Individual scores computed for crit 'importance'
May 26 20:23:50 tournesol bash[21337]: Computing scaling for 2940 non_supertrusted, based on 72 supertrusted
May 26 20:24:35 tournesol bash[21337]: Mehestan for poll 'videos': scores computed for crit 'importance' and mode 'default'
May 26 20:24:38 tournesol bash[21337]: Mehestan for poll 'videos': scores computed for crit 'importance' and mode 'all_equal'
May 26 20:24:40 tournesol bash[21337]: Mehestan for poll 'videos': scores computed for crit 'importance' and mode 'trusted_only'
May 26 20:24:40 tournesol bash[21337]: Mehestan for poll 'videos': done with crit 'importance'
May 26 20:24:40 tournesol bash[21337]: Mehestan for poll 'videos': computing scores for crit 'layman_friendly'
May 26 20:24:45 tournesol bash[21337]: Individual scores computed for crit 'layman_friendly'
May 26 20:24:59 tournesol bash[21337]: Computing scaling for 2940 non_supertrusted, based on 72 supertrusted
May 26 20:25:42 tournesol bash[21337]: Mehestan for poll 'videos': scores computed for crit 'layman_friendly' and mode 'default'
May 26 20:25:45 tournesol bash[21337]: Mehestan for poll 'videos': scores computed for crit 'layman_friendly' and mode 'all_equal'
May 26 20:25:47 tournesol bash[21337]: Mehestan for poll 'videos': scores computed for crit 'layman_friendly' and mode 'trusted_only'
May 26 20:25:48 tournesol bash[21337]: Mehestan for poll 'videos': done with crit 'layman_friendly'
May 26 20:25:48 tournesol bash[21337]: Mehestan for poll 'videos': computing scores for crit 'entertaining_relaxing'
May 26 20:25:52 tournesol bash[21337]: Individual scores computed for crit 'entertaining_relaxing'
May 26 20:26:06 tournesol bash[21337]: Computing scaling for 2940 non_supertrusted, based on 72 supertrusted
May 26 20:26:51 tournesol bash[21337]: Mehestan for poll 'videos': scores computed for crit 'entertaining_relaxing' and mode 'default'
May 26 20:26:54 tournesol bash[21337]: Mehestan for poll 'videos': scores computed for crit 'entertaining_relaxing' and mode 'all_equal'
May 26 20:26:56 tournesol bash[21337]: Mehestan for poll 'videos': scores computed for crit 'entertaining_relaxing' and mode 'trusted_only'
May 26 20:26:57 tournesol bash[21337]: Mehestan for poll 'videos': done with crit 'entertaining_relaxing'
May 26 20:26:57 tournesol bash[21337]: Mehestan for poll 'videos': computing scores for crit 'engaging'
May 26 20:27:03 tournesol bash[21337]: Individual scores computed for crit 'engaging'
May 26 20:27:19 tournesol bash[21337]: Computing scaling for 2940 non_supertrusted, based on 72 supertrusted
May 26 20:28:10 tournesol bash[21337]: Mehestan for poll 'videos': scores computed for crit 'engaging' and mode 'default'
May 26 20:28:13 tournesol bash[21337]: Mehestan for poll 'videos': scores computed for crit 'engaging' and mode 'all_equal'
May 26 20:28:16 tournesol bash[21337]: Mehestan for poll 'videos': scores computed for crit 'engaging' and mode 'trusted_only'
May 26 20:28:16 tournesol bash[21337]: Mehestan for poll 'videos': done with crit 'engaging'
May 26 20:28:16 tournesol bash[21337]: Mehestan for poll 'videos': computing scores for crit 'diversity_inclusion'
May 26 20:28:21 tournesol bash[21337]: Individual scores computed for crit 'diversity_inclusion'
May 26 20:28:37 tournesol bash[21337]: Computing scaling for 2940 non_supertrusted, based on 72 supertrusted
May 26 20:29:20 tournesol bash[21337]: Mehestan for poll 'videos': scores computed for crit 'diversity_inclusion' and mode 'default'
May 26 20:29:23 tournesol bash[21337]: Mehestan for poll 'videos': scores computed for crit 'diversity_inclusion' and mode 'all_equal'
May 26 20:29:25 tournesol bash[21337]: Mehestan for poll 'videos': scores computed for crit 'diversity_inclusion' and mode 'trusted_only'
May 26 20:29:26 tournesol bash[21337]: Mehestan for poll 'videos': done with crit 'diversity_inclusion'
May 26 20:29:26 tournesol bash[21337]: Mehestan for poll 'videos': computing scores for crit 'better_habits'
May 26 20:29:30 tournesol bash[21337]: Individual scores computed for crit 'better_habits'
May 26 20:29:46 tournesol bash[21337]: Computing scaling for 2940 non_supertrusted, based on 72 supertrusted
May 26 20:30:35 tournesol bash[21337]: Mehestan for poll 'videos': scores computed for crit 'better_habits' and mode 'default'
May 26 20:30:39 tournesol bash[21337]: Mehestan for poll 'videos': scores computed for crit 'better_habits' and mode 'all_equal'
May 26 20:30:41 tournesol bash[21337]: Mehestan for poll 'videos': scores computed for crit 'better_habits' and mode 'trusted_only'
May 26 20:30:41 tournesol bash[21337]: Mehestan for poll 'videos': done with crit 'better_habits'
May 26 20:30:41 tournesol bash[21337]: Mehestan for poll 'videos': computing scores for crit 'backfire_risk'
May 26 20:30:46 tournesol bash[21337]: Individual scores computed for crit 'backfire_risk'
May 26 20:31:02 tournesol bash[21337]: Computing scaling for 2940 non_supertrusted, based on 72 supertrusted
May 26 20:31:47 tournesol bash[21337]: Mehestan for poll 'videos': scores computed for crit 'backfire_risk' and mode 'default'
May 26 20:31:50 tournesol bash[21337]: Mehestan for poll 'videos': scores computed for crit 'backfire_risk' and mode 'all_equal'
May 26 20:31:52 tournesol bash[21337]: Mehestan for poll 'videos': scores computed for crit 'backfire_risk' and mode 'trusted_only'
May 26 20:31:53 tournesol bash[21337]: Mehestan for poll 'videos': done with crit 'backfire_risk'
May 26 20:31:57 tournesol bash[21329]: Mehestan for poll 'videos': Done
May 26 20:31:58 tournesol systemd[1]: ml-train.service: Succeeded.
May 26 20:31:58 tournesol systemd[1]: Finished Train ML.
May 26 20:31:58 tournesol systemd[1]: ml-train.service: Consumed 12min 6.582s CPU time.
```
</details>



WIth the current parameters, tournesol scores (= score related to the main criterion "largely_recommended") are between -0.13 and +0.22, with the following distribution:
![Figure_1](https://user-images.githubusercontent.com/4726554/170585281-732390c8-6ff8-4b3d-9647-1c47a5bc1cfc.png)

Some specific thresholds are visible, due to how the regularized median limit the influence of individual contributors:
 - scores for videos rated by 1 (trusted) contributor are bounded inside [-0.025, 0.025]
 - scores for videos rated by 2 (trusted) contributors are bounded inside [-0.05, -0.05], etc.
 
where `0.025 = 1 / (2W)` (to be confirmed with @lenhoanglnh, is this the expected behavior?)

## Recommendations (before/after)

<details>
<summary>Click to see recommendations for "all time"</summary>

|Licchavi<br/>(aggregation of all criteria)|Mehestan<br/>("largely recommended" only)|
|-|-|
![tournesol app_recommendations_date=Month language= (4)](https://user-images.githubusercontent.com/4726554/170587546-817855a0-b87b-4a15-a24a-c0282954162a.png)|![tournesol-vm_recommendations_date= language= largely_recommended=100 reliability=0 pedagogy=0 importance=0 layman_friendly=0 entertaining_relaxing=0 engaging=0 diversity_inclusion=0 better_habits=0 backfire_risk=0 (6)](https://user-images.githubusercontent.com/4726554/170587749-bce542fa-9042-47f3-8058-f0ba4e6d0e9f.png)

</details>

<details>
<summary>Click to see recommendations for "last month"</summary>

|Licchavi<br/>(aggregation of all criteria)|Mehestan<br/>("largely recommended" only)|
|-|-|
|![tournesol app_recommendations_date=Month language= (3)](https://user-images.githubusercontent.com/4726554/170588000-84fac186-0ba9-4e2a-b90b-9cbf21fc9dcc.png)|![tournesol-vm_recommendations_date= language= largely_recommended=100 reliability=0 pedagogy=0 importance=0 layman_friendly=0 entertaining_relaxing=0 engaging=0 diversity_inclusion=0 better_habits=0 backfire_risk=0 (7)](https://user-images.githubusercontent.com/4726554/170588076-b03745f7-408d-48c8-b607-b2023ed50a1b.png)

</details>

## Remaining questions

1. How to scale the tournesol score (and other criteria scores?) based on Mehestan outputs?  
Ideally this transformation (if any) should be applied **before** the scores are saved into the database. In order to use the db as the single source of truth for these scores.

2. If the scores are scale differently (compared to the current Licchavi output), what are the UI components that need to be adapted?
   - entity card (decimals in the tournesol score, etc.)
   - graphs (axes bounds?)
   - browser extension (score format)
   - something else?

**Comment by lfaucon** [comment]
Your comparisons seem to mostly differ due to the all criteria / main criteria. There seems to be a lot less difference between the algorithms if you compare them both with the main criteria ranking.

> How to scale the tournesol score (and other criteria scores?) based on Mehestan outputs?

We can easily find `a` and `b` such that after transforming all scores `x => a * x + b` we get that 5% of the videos are rated above 90, 5% are rated below -90 and 90% of the videos are between 90 and -90. Optionally clip between -100 and 100.

I propose to use percentile because anything that relies on a single video (or max/min) may be very noisy.

Unfortunately, I don't think that at this stage we have a robust method to correctly shift the score=0 such that more than half of the video have a chance to get a positive score. This would be helped later with the Like/Dislike feature.

I propose also to compute the Poll Scaling only based on the main criteria and apply the same scaling to all criterias. This way we would be able to observe whether some criterias are more or less spread than others.

Now that I think slightly more about this proposition, we may have the issue that the 1-rating threshold is what is selected for the precentiles? This may be fixed by scaling only from entities with a minimum number of contributors (e.g. ~4).

@aidanjungo @lenhoanglnh What do you think of this approach for now?

> something else?

I think you covered everything. I checked the twitter bot. It doesn't display scores.

**Comment by lfaucon** [comment]
Other thing to look at would be the scale of the individual scores. The current algorithm makes the score extremely close to aggregated score (which makes no sense), but what about Mehestan? Do we see more differences? Do aggregated and individual scores seems to be on the same scale? There is the concept of scaling for an inidividual. Based on my understanding, it may make more sense to display to a contributor his/her scaled scores

**Comment by lfaucon** [comment]
A lot of Science4All in the all time top scores... Are we worried that we may look biased?

Maybe we could joke about it for promotion "Come vote if you don't think Lê's videos should be at the top of Tournesol's recommendations"

**Comment by lenhoanglnh** [comment]
> @aidanjungo @lenhoanglnh What do you think of this approach for now?

I propose the following fix: find ```a``` and ```b``` such that the Mehestan-normalized contributors' scores verify the percentile conditions you consider.
Then rescale all contributors' scores into ```a*x + b``` (maybe call it the "quantile-Mehestan-normalized scores").
And then apply the last step of Mehestan to the quantile-Mehestan-normalized scores (instead of the Mehestan-normalized scores) to output global scores.

One important advantage of doing so is that "0" will still represent the default value of unscored videos.

**Comment by amatissart** [comment]
Decision:
 * after a first "dry run" of Mehestan, compute `a` and `b` such as the 5th and 95th percentiles of scores are -90 and 90. A second run will be used to compute the final scores, after individual scores are scaled with `a*x+b`.
 
 * the personal scores returned by the API should be scales accordingly, to be comparable with global scores.

**Comment by amatissart** [comment]
Updated scores on recent production data after applying the scaling described above  
(implemented in #1005, run on commit 92e616f)

### Distribution

![Figure_1](https://user-images.githubusercontent.com/4726554/174835968-abaad67c-a234-4254-b4ea-8686e748a785.png)

<details>
<summary>Click for distribution details</summary>

```
mean       -7.400982
std        80.544131
min      -438.415360
5%        -86.872189
10%       -83.229292
15%       -78.708966
20%       -70.723132
25%       -59.986535
30%       -48.852610
35%       -37.089778
40%       -27.980770
45%       -18.294185
50%       -10.138298
55%        -3.404447
60%         3.804517
65%        10.911559
70%        19.814852
75%        32.051675
80%        44.184307
85%        60.419702
90%        77.932194
95%        92.961090
max       709.995148
```

</details>

### Recommendations (before/after)
Compared to Licchavi (current production), **sorted by "largely recommended" only**

<details>
<summary>Click to see recommendations for "all time" :sunflower:</summary>

|Licchavi<br>sorted by "largely recommended"|Mehestan<br>with poll scaling applied<br>|
|-|-|
|![tournesol app_recommendations_date= language= largely_recommended=100 reliability=0 pedagogy=0 importance=0 layman_friendly=0 entertaining_relaxing=0 engaging=0 diversity_inclusion=0 better_habits=0 backfire_risk=0 (2)](https://user-images.githubusercontent.com/4726554/174837555-b325608e-9c66-4102-a743-1eae28137367.png)|![tournesol-vm_recommendations_date=Month language= largely_recommended=100 reliability=0 pedagogy=0 importance=0 layman_friendly=0 entertaining_relaxing=0 engaging=0 diversity_inclusion=0 better_habits=0 backfire_risk=0 offset=0 (7)](https://user-images.githubusercontent.com/4726554/174837682-efbf143b-1f06-4bfc-8f57-fec982e3b1f4.png)|
</details> 

<details>
<summary>Click to see recommendations for "last month" :sunflower:</summary>

|Licchavi<br>sorted by "largely recommended"|Mehestan<br>with poll scaling applied<br>|
|-|-|
|![tournesol app_recommendations_date= language= largely_recommended=100 reliability=0 pedagogy=0 importance=0 layman_friendly=0 entertaining_relaxing=0 engaging=0 diversity_inclusion=0 better_habits=0 backfire_risk=0 (3)](https://user-images.githubusercontent.com/4726554/174838593-2ed0be13-1b1c-4f5f-9e30-8c2a430553d0.png)|![tournesol-vm_recommendations_date=Month language= largely_recommended=100 reliability=0 pedagogy=0 importance=0 layman_friendly=0 entertaining_relaxing=0 engaging=0 diversity_inclusion=0 better_habits=0 backfire_risk=0 offset=0 (6)](https://user-images.githubusercontent.com/4726554/174838636-f2e14ee3-7d49-42e9-a630-fba8c846cef9.png)

</details>

**Comment by amatissart** [comment]
As tou can see, if the intention was to scale scores such as Tournesol scores for top videos is ~ 100, then +90 for the 95th percentile is too high.
